contents/javascript/Javascript-promises.md

```diff
@@ -90,57 +86,130 @@ These situations correspond to the 3 states of Promises in Javascript.
+
+```javascript
+function intensewait(val) {
+    // return val after a short wait
+    return new promise(resolve => {
+        settimeout(function(){
+            resolve(val);
         }, 1000);
     });
 }
 
-update().then(result => console.log(result));
+/** note these funcs now return a promise **/
+function getdata() {
+    return intensewait('some random data');
+}
 
+function filterdata(data, callback) {
+    return intensewait(data.split(' '));
+}
+
+function main() {
+    return getdata()
+        .then(data => filterdata(data))
+        .then(fltered => console.log(filtered));
+}
+
+main();
 ```
 
-The first step of creating a Promise involves invoking the `Promise` constructor which takes in just 1 parameter known as the executor function. This is the function that will be executed by the Promise. This function has two parameters: resolve, reject. In our example, we only use the resolve function to indicate that our promise has executed successfully with the result `x`. The reject function may be used to indicate that some error has occured during execution. 
+Using Promises helps your code to be clean and easy to read, which makes it easy for anyone to instantly know what the code does, instead of having to trace through the code written with callbacks.
 
-After defining the function, we call it, and we call another function `.then()` upon the result. `.then()` is used to consume the result from a fulfilled Promise and act on it. In this case, we are using it to log our result. More importantly, it lets us chain promises to avoid the deep nesting known as callback hell.
+### The iterative style of promises
 
-Here's the callback hell code rewritten with Promises.
+The "promise style", with all the `.then()` to pass data from one function to the next, it is kinda like functional programming. The original promise is passed from one `.then()` to the other, and with each `.then()`, a new promise is returned for the next `.then()` to work on.
 
-```javascript
+There is probably a small group of you out there is the confused by the whole block of text above. And that is exactly why they have the "async await" style of using promises. Well, callbacks aint exactly for everybody.
 
-function update(numTimes) {
-    return new Promise((resolve, reject) => {
-        var x = "I have not been updated";
-        setTimeout(function() {
-            x = "I have been updated " + numTimes + " times.";
-            console.log(x);
-            resolve(numTimes);
+The `async` keyword ensures that the function returns a promise, and as for `await`, you can think of it as waiting for some promise to return the value before continuing. So back to the same example.
+
+So here is the code in the "promise style".
+
+```javascript
```

Comment:
Hmm why repeat this code snippet again? Btw, can you camelcase the function names?